src/rust/include/tracing.h

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2020 Jack Grigg
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef TRACING_INCLUDE_H_
+#define TRACING_INCLUDE_H_
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+#include <memory>
+
+extern "C" {
+#endif
+#ifdef WIN32
+typedef uint16_t codeunit;
+#else
+typedef uint8_t codeunit;
+#endif
+
+struct TracingHandle;
+typedef struct TracingHandle TracingHandle;
+
+/// Initializes the tracing crate, returning a handle for the logging
+/// component. The handle must be freed to close the logging component.
+///
+/// If log_path is NULL, logging is sent to standard output.
+TracingHandle* tracing_init(
+    const codeunit* log_path,
+    size_t log_path_len,
+    const char* initial_filter,
+    bool log_timestamps);
+
+/// Frees a tracing handle returned from `tracing_init`;
+void tracing_free(TracingHandle* handle);
+
+/// Reloads the tracing filter.
+void tracing_reload(TracingHandle* handle, const char* new_filter);
+
+struct TracingCallsite;
+typedef struct TracingCallsite TracingCallsite;
+
+/// Creates a tracing callsite.
+///
+/// You should usually call the `TracingLog` macro (or one of the helper
+/// macros such as `TracingInfo`) instead of calling this directly.
+///
+/// This MUST ONLY be called to assign a `static void*`, and all string
+/// arguments MUST be `static const char*`.
+TracingCallsite* tracing_callsite(
+    const char* name,
+    const char* target,
+    const char* level,
+    const char* file,
+    uint32_t line,
+    const char** fields,
+    size_t fields_len,
+    bool is_span);
+
+struct TracingSpanHandle;
+typedef struct TracingSpanHandle TracingSpanHandle;
+
+struct TracingSpanGuard;
+typedef struct TracingSpanGuard TracingSpanGuard;
+
+/// Creates a span for a callsite.
+///
+/// The span must be freed when it goes out of scope.
+TracingSpanHandle* tracing_span_create(const TracingCallsite* callsite);
+
+/// Clones the given span.
+///
+/// Both spans need to be separately freed when they goe out of scope.
```

Comment:
```suggestion
/// Both spans need to be separately freed when they go out of scope.
```

Reply:
Heh, you're always so quick to the punch 😄